src/v/cloud_storage/partition_manifest.cc

```diff
@@ -285,42 +285,43 @@ partition_manifest::get_start_kafka_offset() const {
 partition_manifest::const_iterator
 partition_manifest::segment_containing(kafka::offset o) const {
     vlog(cst_log.debug, "Metadata lookup using kafka offset {}", o);
+    if (_segments.empty()) {
+        return end();
+    }
     // Kafka offset is always <= log offset.
     // To find a segment by its kafka offset we can simply query
     // manifest by log offset and then traverse forward until we
-    // will find matching segment.
-    auto it = segment_containing(kafka::offset_cast(o));
-    if (it == end()) {
-        return it;
-    }
-    auto prev = end();
+    // find a matching segment.
+    auto it = _segments.lower_bound(kafka::offset_cast(o));
+    // We need to find first element which has greater kafka offset than
+    // the target and step back. It is possible to have a segment that
+    // doesn't have data batches. This scan has to skip segments like that.
     while (it != end()) {
-        auto base = it->second.base_offset - it->second.delta_offset;
-        if (base > o) {
-            // We need to find first element which has greater kafka
-            // offset then the target and step back. It is possible
-            // to have a segment that doesn't have data batches. This
-            // scan has to skip segments like that.
-            break;
+        if (it->second.base_kafka_offset() > o) {
+            // The beginning of the manifest already has a base offset that
+            // doesn't satisfy the query.
+            if (it == begin()) {
+                return end();
+            }
+            // On the first segment we see with a base kafka offset higher than
+            // 'o', return its previous segment.
+            return std::prev(it);
```

Comment:
In case if the fetch request has offset below the `start_offset` of the partition the code will return the first segment in the manifest. I'm not sure that this behavior is correct. Maybe we should return `end()` in this case and only return `std::prev(it)` if previously we checked that segment and found that its base offset is smaller than the request which will mean that it actually contains the offset we want.

Reply:
Yeah, maybe I can make it more explicit by checking `_start_offset` up front, but that case is handled, unless I'm misunderstanding the question.